File: app/src/main/java/com/example/android_demo_application/fragments/LoginFragment.java
package com.example.android_demo_application.fragments;

import android.app.Application;
import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.example.android_demo_application.MyApplication;
import com.example.android_demo_application.R;
import com.example.android_demo_application.utils.HttpUtils;
import com.example.android_demo_application.utils.SharedPreferenceUtils;
import com.google.android.material.textfield.TextInputEditText;

public class LoginFragment extends Fragment {
    TextInputEditText yonghuming;
    TextInputEditText mima;
    Button button;
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_login,container,false);
        yonghuming = view.findViewById(R.id.text_intput_username);
        mima = view.findViewById(R.id.text_input_password);
        button = view.findViewById(R.id.denglu);
        Handler handler = new Handler(){
            @Override
            public void handleMessage(@NonNull Message msg) {
                //主线程上

                String s = (String) msg.obj;
                if(s.equals("")){
                    //保存 登录逻辑
                    SharedPreferenceUtils.saveUserInfo(yonghuming.getText().toString(),mima.getText().toString());
                    MyApplication.setIsLoggedIn(true);
                    MyApplication.setUserName(yonghuming.getText().toString());
                    getActivity().finish();
                    return;
                }
                MyApplication.setIsLoggedIn(false);
                MyApplication.setUserName("");
                    SharedPreferenceUtils.empty();
                    Toast.makeText(MyApplication.getContext(),s,Toast.LENGTH_SHORT).show();
            }
        };
        button.setOnClickListener(v ->
                MyApplication.getPools().execute(() -> {
                    String string = HttpUtils.login(yonghuming.getText().toString(),mima.getText().toString());
                    Message message = Message.obtain();
                    message.setTarget(handler);
                    message.obj = string;
                    handler.sendMessage(message);
                }
                )
        );
        return view;
    }
}

File: app/src/main/java/com/example/android_demo_application/fragments/LoginFragment.kt
package com.example.android_demo_application.fragments

import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.Toast
import androidx.fragment.app.Fragment
import com.example.android_demo_application.MyApplication
import com.example.android_demo_application.R
import com.example.android_demo_application.presenter.LoginPresenter
import com.example.android_demo_application.utils.HttpUtils.login
import com.example.android_demo_application.utils.SharedPreferenceUtils
import com.google.android.material.textfield.TextInputEditText
import java.lang.ref.WeakReference

class LoginFragment : Fragment(),LoginView {
    private val presenter:LoginPresenter by lazy { LoginPresenter() }
    private lateinit var itemView:View
    var yonghuming: TextInputEditText? = null
    var mima: TextInputEditText? = null
    var button: Button? = null
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        presenter.attachView(this)
        itemView = inflater.inflate(R.layout.fragment_login, container, false)
        init()
        return itemView
    }
    private fun init(){
        yonghuming = itemView.findViewById(R.id.text_intput_username)
        mima = itemView.findViewById(R.id.text_input_password)
        button = itemView.findViewById(R.id.denglu)
        button?.setOnClickListener{
            presenter.login(yonghuming?.text.toString(),mima?.text.toString())
        }
    }
    override fun onDestroyView() {
        super.onDestroyView()
        presenter.detachView()
    }

    override fun onSuccess() {
        SharedPreferenceUtils.saveUserInfo(yonghuming!!.text.toString(), mima!!.text.toString())
        MyApplication.setIsLoggedIn(true)
        MyApplication.userName = yonghuming!!.text.toString()

        //发送广播
        var intent = Intent("login")
        intent.setPackage(activity?.packageName)
        activity?.sendBroadcast(intent)
        activity?.finish()
    }

    override fun onFail(s: String?) {
        MyApplication.setIsLoggedIn(false)
        MyApplication.userName = ""
        //清空 SharedPreference
        SharedPreferenceUtils.empty()
        Toast.makeText(MyApplication.context, s, Toast.LENGTH_SHORT).show()
    }
}